packages/wallet/src/redux/protocols/virtual-defunding/reducer.ts

```diff
@@ -0,0 +1,188 @@
+): ProtocolStateWithSharedData<states.VirtualDefundingState> {
+  if (routesToConsensusUpdate(action, protocolState.protocolLocator)) {
+    let jointChannel: ConsensusUpdateState;
+    ({ protocolState: jointChannel, sharedData } = consensusUpdateReducer(
+      protocolState.jointChannel,
+      sharedData,
+      action,
+    ));
+    switch (jointChannel.type) {
+      case 'ConsensusUpdate.Failure':
+        return { protocolState: states.failure({}), sharedData };
+      case 'ConsensusUpdate.Success':
+        const {
+          hubAddress,
+          ledgerChannelId,
+          targetChannelId: appChannelId,
+          ourIndex,
+          processId,
+        } = protocolState;
+        // TODO: We probably need to start this earlier to deal with commitments coming in early
+
+        const latestAppCommitment = getLatestCommitment(appChannelId, sharedData);
+
+        const proposedAllocation = [
+          latestAppCommitment.allocation[ourIndex],
+          latestAppCommitment.allocation[1 - ourIndex],
+        ];
+        const proposedDestination = [latestAppCommitment.destination[ourIndex], hubAddress];
+        let ledgerChannel: ConsensusUpdateState;
+        ({ protocolState: ledgerChannel, sharedData } = initializeConsensusUpdate({
+          processId,
+          protocolLocator: makeLocator(
+            protocolState.protocolLocator,
+            CONSENSUS_UPDATE_PROTOCOL_LOCATOR,
+          ),
+          channelId: ledgerChannelId,
+          proposedAllocation,
+          proposedDestination,
+          clearedToSend: true,
+          sharedData,
+        }));
+
+        return {
+          protocolState: states.waitForLedgerChannelUpdate({ ...protocolState, ledgerChannel }),
+          sharedData,
+        };
+      default:
+        return {
+          protocolState: { ...protocolState, jointChannel },
+          sharedData,
+        };
+    }
+  }
+  return { protocolState, sharedData };
+}
+
+function waitForLedgerChannelUpdateReducer(
+  protocolState: states.WaitForLedgerChannelUpdate,
+  sharedData: SharedData,
+  action: VirtualDefundingAction,
+): ProtocolStateWithSharedData<states.VirtualDefundingState> {
+  if (routesToConsensusUpdate(action, protocolState.protocolLocator)) {
+    let ledgerChannel: ConsensusUpdateState;
+    ({ protocolState: ledgerChannel, sharedData } = consensusUpdateReducer(
+      protocolState.ledgerChannel,
+      sharedData,
+      action,
+    ));
+    switch (ledgerChannel.type) {
+      case 'ConsensusUpdate.Failure':
+        return { protocolState: states.failure({}), sharedData };
+      case 'ConsensusUpdate.Success':
+        return { protocolState: states.success({}), sharedData };
+      default:
+        return {
+          protocolState: { ...protocolState, ledgerChannel },
+          sharedData,
+        };
+    }
+  }
+  return { protocolState, sharedData };
+}
+
+function getLedgerChannelId(jointChannelId: string, sharedData: SharedData): string {
```

Comment:
This name seems a bit ambiguous, because the joint channel whose ID you're passing in is already a ledger channel, but it's not the right ledger channel. Is there a name that makes it clear which ledger channel you're after?

Reply:
This function gets refactored in another PR and has a more sensible name.